src/mgr/DaemonServer.cc

```diff
@@ -841,12 +894,59 @@ bool DaemonServer::handle_command(MCommand *m)
   }
 }
 
+bool DaemonServer::_prune_pending_service_map()
+{
+  utime_t cutoff = ceph_clock_now();
+  cutoff -= g_conf->mgr_service_beacon_grace;
+  bool changed = false;
+  auto p = pending_service_map.services.begin();
+  while (p != pending_service_map.services.end()) {
+    auto q = p->second.daemons.begin();
+    while (q != p->second.daemons.end()) {
+      auto d = service_state.get_daemon(p->first, q->first);
+      if (d->last_beacon == utime_t()) {
+	// we must have just restarted; assume they are alive now.
+	d->last_beacon = ceph_clock_now();
+	++q;
+	continue;
+      }
+      if (d->last_beacon < cutoff) {
+	dout(10) << "pruning stale " << p->first << "." << q->first
+		 << " last_beacon " << d->last_beacon << dendl;
+	q = p->second.daemons.erase(q);
+	changed = true;
+      } else {
+	++q;
+      }
+    }
+    if (p->second.daemons.empty()) {
+      p = pending_service_map.services.erase(p);
+      changed = true;
+    } else {
+      ++p;
+    }
+  }
+  return changed;
```

Comment:
Maybe set pending_service_map_dirty to `changed` so that caller can just check that one thing instead of that and the return value.